File: apps/backend/app/models/audit.py
"""
Audit Log Model
Tracks all user actions for security and compliance
"""

from datetime import datetime
from typing import Optional
from sqlmodel import SQLModel, Field, JSON, Column
from sqlalchemy import Text


class AuditLog(SQLModel, table=True):
    """Audit log for tracking user actions"""
    __tablename__ = "audit_logs"
    
    id: Optional[int] = Field(default=None, primary_key=True)
    user_id: int = Field(foreign_key="users.id", index=True)
    action: str = Field(max_length=100, index=True)  # e.g., "create", "update", "delete", "login"
    resource_type: str = Field(max_length=50, index=True)  # e.g., "flow", "user", "bot"
    resource_id: Optional[int] = Field(default=None, index=True)
    details: dict = Field(default={}, sa_column=Column(JSON))  # Additional context
    ip_address: Optional[str] = Field(default=None, max_length=45)
    user_agent: Optional[str] = Field(default=None, sa_column=Column(Text))
    status: str = Field(default="success", max_length=20)  # success, failed, error
    error_message: Optional[str] = Field(default=None, sa_column=Column(Text))
    created_at: datetime = Field(default_factory=datetime.utcnow, index=True)
    
    class Config:
        json_schema_extra = {
            "example": {
                "user_id": 1,
                "action": "create",
                "resource_type": "flow",
                "resource_id": 123,
                "details": {"name": "My Workflow", "status": "draft"},
                "ip_address": "192.168.1.1",
                "user_agent": "Mozilla/5.0...",
                "status": "success"
            }
        }

File: apps/backend/src/audit/domain/audit.ts
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';

/**
 * AuditLog domain entity - theo schema má»›i
 * Table: audit_logs
 */
export class AuditLog {
  @ApiProperty({ type: String })
  id: string;

  @ApiProperty({ type: String })
  userId: string;

  @ApiProperty({ type: String })
  workspaceId: string;

  @ApiProperty({ type: String, example: 'create' })
  action: string;

  @ApiProperty({ type: String, example: 'bot' })
  resourceType: string;

  @ApiProperty({ type: String })
  resourceId: string;

  @ApiPropertyOptional({ type: Object })
  details?: Record<string, any>;

  @ApiPropertyOptional({ type: String })
  ipAddress?: string | null;

  @ApiPropertyOptional({ type: String })
  userAgent?: string | null;

  @ApiProperty()
  createdAt: Date;
}

/**
 * DataAccessLog domain entity - theo schema má»›i
 * Table: data_access_logs
 */
export class DataAccessLog {
  @ApiProperty({ type: String })
  id: string;

  @ApiProperty({ type: String })
  userId: string;

  @ApiProperty({ type: String })
  workspaceId: string;

  @ApiProperty({ type: String })
  tableName: string;

  @ApiProperty({ type: String })
  recordId: string;

  @ApiProperty({ type: String, enum: ['read', 'write', 'delete'] })
  action: 'read' | 'write' | 'delete';

  @ApiProperty()
  createdAt: Date;
}
